llvm/lib/ProfileData/SampleProfReader.cpp

```diff
@@ -223,6 +223,11 @@ static bool ParseLine(const StringRef &Input, LineType &LineTy, uint32_t &Depth,
 
   if (Input[Depth] == '!') {
     LineTy = LineType::Metadata;
+    // This metadata is only for manual inspection only. We already created a
+    // FunctionSamples and put it in the profile map, so there is no point
+    // to skip profiles even they have no use for ThinLTO.
+    if (Input == StringRef(" !Flat"))
```

Comment:
I think we need to make sure that profiles in different formats, originating from the same source, to be equivalent. This means that if we do a round-trip conversion—from a "text" profile to an "extbinary" profile and back to a "text" profile—the two text profiles before and after the conversion should be identical. However, it seems that here the "!flat" metadata would be lost because we treat "!Flat" as a no-op, and the extended binary format won't be set to split-layout.

Reply:
Text and Extbinary mode was already not perfectly compatible. There are already various types of info that will be lost when converting Extbinary to Text, for example Profile Symbol List, flags such as PartialProfile, ProfileIsFS etc. Furthermore, the sample profile writer splits flat profiles from non-flat profiles at runtime, and the reader ignores the flag if used by llvm-profdata (instead of used by thinLTO), so the flag is a no-op in this sense, and the round trip output will be identical, unless the user intentionally marked a non-flat function as flat, which doesn't make sense anyways (profile reader could emit a warning in this case but it is also pointless, because 1. in llvm-profdata the writer will correct this, 2. in thinLTO the flat section is skipped intentionally so we don't even know there's non-flat function inside 3. in other compilation mode we read all functions with matching names regardless they are flat or not.)

Reply:
Does losing !flat change the semantics of the profile? If not, they are still equivalent?  On other hand, instead of using meta data, would a comment line be sufficient?

Reply:
Makes sense, thanks for the clarification. I thought for other incompatible cases( Profile Symbol List, PartialProfile, ProfileIsFS, etc), mostly because they don't support text format, but here as the text format is added, so maybe it's not hard to support the compatibility.